E2e tests for covering the main functionalities

amandassa opened the pull request:
Fixes #7801 

#### Short description of what this resolves:
The Selenium e2e tests are insufficient and create huge overhead on test maintenance.

#### Changes proposed in this pull request:
Introduce true end-to-end tests to validate the most important functionalities using testRigor opensource option.

#### Does your code include anything generated by an AI Engine? No

adunsulag commented:
@amandassa can you tell us a bit more about test rigor.  Did you create an account for OpenEMR?  Or will we need to setup our own account?  Without us having to dig through all of your documentation I'd like to hear the pros in terms of both immediate impact and long term maintainability of this solution.  

How does testRigor handle the hallucination effects of LLM models and need for 100% reliability that the test execution context will function the exact same on each test execution?

What are the limits on the public / open source offering for use with our CI engine?  

In our CI process we run through more than a dozen environments for every single PR.  We'd need to understand the throughput and limitations of this 3rd party piece before we make our CI engine dependent on it.

The overall tests look nice and our lack of familiarity with the tool makes it difficult to evaluate this PR from a correctness point of view so we need to better understand the tool as OpenEMR looks at product time horizons in terms of years and decades rather than months.

amandassa commented:
@adunsulag

**Q: Can you tell us a bit more about test rigor?**
A: testRigor is a platform where end-to-end testing is made simple by using plain English instructions to perform tests from user perspective.

**Q: Did you create an account for OpenEMR? Or will we need to setup our own account?**
A: Yes, I created a Public Open Source account for OpenEMR and a new test suite for it.

I added the Suite ID and CI Integration Token to the test.yml file as workflow environment variables from GitHub Secrets, that way the integration to your CI is made possible without the need for you to set up a new account and suite.

The CI Integration is made possible by using [testRigor command line tool](https://testrigor.com/command-line). The CLI will use values from Suite ID and CI Integration Token to run testRigor test cases from GitHub workflow environment. That way, the test cases from this PR **and** new testRigor test cases added through contributors can be tested using the account I created for OpenEMR.

On addition to the integration, I provided info on how to set up a new account and suite for OpenEMR in the testRigor/README.md file for developers that intend to run testing in a local OpenEMR environment. By following that setup, any contributor is able to create new testRigor test cases for OpenEMR.

**Q: How does testRigor handle the hallucination effects of LLM models and need for 100% reliability that the test execution context will function the exact same on each test execution?**
A: The specific test steps are parsed and executed emulating how a human would do it, without AI. AI is used to build steps if an unrecognized phrase is provided.

**Q: What are the limits on the public / open source offering for use with our CI engine?**
A: The testRigor Public Open Source account is free forever.

Let me know if you have any other questions, and I'll be happy to get back to you.

adunsulag commented:
Hi @amandassa thanks for taking time to answer some of my questions.

> Q: What are the limits on the public / open source offering for use with our CI engine?
A: The testRigor Public Open Source account is free forever.



 I understand the account is free for public offerings but I want to understand the number of parallel tests that can be executed on the public account.  For example if we have 4 PRs that come in and they spin up 12 environments.  We will be executing 48 e2e tests all in parallel.  Its rare, but not unheard of for us to get 4 PRs at the same time, but we've at times had 2 at the same time which would be 24 simultaneous executions.

This blog post from a year ago on testrigor's website says the limit for parallel test execution is one for the public plan.  
```
https://testrigor.com/how-to-articles/how-to-execute-test-cases-in-parallel-in-testrigor/
```
If that's the case it will slow down our e2e testing environment a lot if everything has to be done sequentially for the 12 different environments we work with.  Have those limits changed in the last year?  Otherwise it sounds like we'd need to move to a paid plan in order to have timely CI checks for what we do.

amandassa commented:
@adunsulag 
Hi Stephen, I have talked to testRigor and they are willing to allocate 12 parallels for free for OpenEMR CI.

adunsulag commented:
Hey @amandassa sorry for the delay in getting back to you.  This sounds great, we talked as admins today and we'd like to bring this PR in.  We noticed the tests are failing.  Does testrigor require reaching the docker externally from the testrigor service?  It looks like that would need to be fixed in order to get this brought in.  Once the tests are passing we can merge this in.

amandassa commented:
Hi, the tests are failing because the testrigor command is not installed successfully.
I've been trying to install testrigor.cli into the workflow, but my approaches weren't successful by now.
I already tried the install command at the same file as the bash script for running tests. And also have tried adding the install command at the ci script. I would like to know what approach would be best for you (adding it as a dev dependency maybe?)
Can you help me with that?

raskolnikov-rodion commented:
Hey, it's been a while since I last contributed here. Hope everyone is doing well! 🙂

This PR caught my eye, so here are a couple of comments:

> Hi, the tests are failing because the testrigor command is not installed successfully

@amandassa have you tried adding it to the dev dependencies? Then you can run in the command line using `npx`, like this:

`npx testrigor test-suite run ...`

Another thing I noticed is that we have some tokens and ids in the `run_testrigor_tests.sh` file. I don't know exactly how these are used by testRigor, but if these values are associated with a specific OpenEMR account, might be better to store them using [Secrets](https://docs.github.com/en/actions/security-for-github-actions/security-guides/using-secrets-in-github-actions).

I haven't used testRigor yet and it does look cool. However, if the idea is to add some E2E tests that interact with the UI, have you considered exploring other tools that are popular in the frontend engineering space, such as Cypress or Playwright?

I'm not saying I'm against using testRigor (or any other AI-based tool), just that it would be good to explore other tools and understand the trade-offs to make sure the tool is right for the project. Though if that discussion has already happened, then please carry on, I'm just landing here!

Good job taking the initiative to add e2e tests. Hope you have fun contributing to this project

amandassa commented:
Hi @raskolnikov-rodion . Thank you for contributing to the discussion, I appreciate it.
- I have added it to the dependency install on "prepare" script now. When adding to "devDependencies" the installation weren't found. I couldn't figure why.
- I have replaced the token with a secret. I would like very much to talk to a mainteiner in order to help them save the token as a secret before this is merged.

The current blocker I came across is that among the 12 tests, I have 2 specific tests for the Patient Portal. However, I can't access it the same way as the admin page. Does the installation of the portal have to be added to the builds in order for it to be reached?

raskolnikov-rodion commented:
Hi @amandassa 

> I have added it to the dependency install on "prepare" script now. When adding to "devDependencies" the installation weren't found. I couldn't figure why.

The problem with adding the package on the devDependencies together with the "prepare" script using the `-g` flag is that now the workflow installs the package twice, both locally and globally. So it's best to do either one or the other, the local installation with devDependencies being preferred, if possible.

I haven't looked into it deeply, but the reason why the installation was not found when adding the package to devDependencies seems to be related to how the dependencies are managed during the workflow. In the `ciLibrary.source` file, the last command in `main_build` is `rm -fr node_modules`. I haven't worked on the setup for the CI so I don't know why it does that, but you might need to reinstall the local dependencies before you run the tests.

Also, I just noticed that the package is marked as "beta". Do you know if there are plans for a non-beta release any time soon? Is this stable enough to be integrated into the project? Looking at the download statistics on npm, adoption seems to be quite low. I'd be a bit concerned about what this would mean in terms of long-term support.

amandassa commented:
Hi I am facing an issue with most of the Apache builds (except for the build PHP 8.2 - Apache - MariaDB 11.4). On those builds the portal is enabled (checked) but I get stuck on this page with no "Create Patient Portal" options.

![image](https://github.com/user-attachments/assets/1848511a-08ec-4eca-aefe-e7e323099ac7)

Two tests are validating the patient portal functionality. One other test validates the appointments functionality. All three fail on only on the mentioned builds because of this issue.

I appreciate any information that can help me progress. Thanks in advance
